app/src/main/java/de/westnordost/streetcomplete/quests/building_entrance/AddEntrance.kt

```diff
@@ -0,0 +1,72 @@
+package de.westnordost.streetcomplete.quests.building_entrance
+
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.elementfilter.toElementFilterExpression
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmElementQuestType
+import de.westnordost.streetcomplete.data.osm.osmquests.Tags
+import de.westnordost.streetcomplete.data.user.achievements.QuestTypeAchievement.PEDESTRIAN
+
+class AddEntrance : OsmElementQuestType<EntranceAnswer> {
+
+    private val withoutEntranceFilter by lazy { """
+        nodes with
+          !entrance and !barrier and noexit != yes
+    """.toElementFilterExpression() }
+
+    private val buildingWaysFilter by lazy { """
+        ways, relations with building and building !~ yes|no|service|shed|house|detached|terrace|semi|semidetached_house
+    """.toElementFilterExpression() }
+
+    private val incomingWaysFilter by lazy { """
+        ways with
+          highway ~ path|footway and area != yes and access !~ private|no
+    """.toElementFilterExpression() }
+
+    private val excludedWaysFilter by lazy { """
+        ways with (tunnel and tunnel != no) or covered=yes
+    """.toElementFilterExpression() }
```

Comment:
You can combine both filters

```suggestion
    private val incomingWaysFilter by lazy { """
        ways with
          highway ~ path|footway
          and area != yes
          and access !~ private|no
          and (!tunnel or tunnel = no)
          and covered != yes
    """.toElementFilterExpression() }
```

Reply:
Actually no, it is not the same.

Note that node may touch OK way and banned way.

I added test for that.

Reply:
Is https://github.com/streetcomplete/StreetComplete/blob/ad891e038076a82651a65b96aa103b14fa75e157/app/src/test/java/de/westnordost/streetcomplete/quests/building_entrance/AddEntranceTest.kt explaining difference?

Node may be both in a banned way and acceptable way.

In particular, https://www.openstreetmap.org/node/3091300654 should not trigger entrance quest

Reply:
right, I understand